### PR TITLE
MEV Sandwich Trades `dex.sandwiches`

### PR DESCRIPTION
**Pull request by hildobby**
Brief comments on the purpose of your changes:

**For Dune Engine V2**

I've checked that:

### General checks:
* [ ] I tested the query on dune.com after compiling the model with dbt compile (compiled queries are written to the target directory)
* [ ] I used "refs" to reference other models in this repo and "sources" to reference raw or decoded tables 
* [ ] if adding a new model, I added a test
* [ ] the filename is unique and ends with .sql
* [ ] each sql file is a select statement and has only one view, table or function defined  
* [ ] column names are `lowercase_snake_cased`
* [ ] if adding a new model, I edited the dbt project YAML file with new directory path for both models and seeds (if applicable)
* [ ] if wanting to expose a model in the UI (Dune data explorer), I added a post-hook in the JINJA config to add metadata (blockchains, sector/project, name and contributor Dune usernames)

### Pricing checks:
* [ ] `coin_id` represents the ID of the coin on coinpaprika.com
* [ ] all the coins are active on coinpaprika.com (please remove inactive ones)

### Join logic:
* [ ] if joining to base table (i.e. ethereum transactions or traces), I looked to make it an inner join if possible

### Incremental logic:
* [ ] I used is_incremental & not is_incremental jinja block filters on both base tables and decoded tables
  * [ ] where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to base table (i.e. ethereum transactions or traces), I applied join condition where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to prices view, I applied join condition where minute >= date_trunc("day", now() - interval '1 week')

### CODE REVIEW

**Comment by Hosuke** [comment]
Dupes produced by checking [blockchain, frontrun_tx_hash, frontrun_index]
https://dune.com/queries/2520420?d=1

**Comment by Hosuke** [comment]
Add `frontrun_taker` to `unique_key` based on observation from test failure results:
https://dune.com/queries/2520509?d=1

**Comment by Hosuke** [comment]
Hi @hildobby,
sorry I could not find an exact `unique_key` to distinguish these records.

**Comment by hildobby** [comment]
> Added `frontrun_taker` to `unique_key` based on observation from test failure results: https://dune.com/queries/2520509?d=1

Fixed it now @Hosuke :)

**Review by Hosuke — APPROVED** [review]
LGTM.✅
Thank you @hildobby 

[Sample dex.sandwiches](https://dune.com/queries/2601380?d=1)

File: models/dex/dex_sandwiches.sql [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+        incremental_strategy = 'merge',
+        unique_key = ['blockchain', 'sandwiched_pool', 'frontrun_tx_hash', 'frontrun_taker', 'frontrun_index', 'currency_address'],
+        post_hook='{{ expose_spells(\'["ethereum", "bnb", "avalanche_c", "gnosis", "optimism", "arbitrum", "fantom", "polygon"]\',
+                                "sector",
+                                "dex",
+                                \'["hildobby"]\') }}'
+        )
+}}
+
+WITH trades AS (
```
Comment (jeff-dude) [inline models/dex/dex_sandwiches.sql:15]: similar to the other recently reviewed PR, can we leverage the for loop approach here, with dictionary variables?

**Review by jeff-dude — CHANGES_REQUESTED** [review]

File: models/dex/dex_sandwiches.sql (hunk 1 above)
Comment (jeff-dude) [reply models/dex/dex_sandwiches.sql:15]: looking closer, now i see the runtime in the ci test, i wonder if we should split this out per blockchain, then rollup as usual. this is a lottt of data for one spell to crunch all these joins and such

Comment (hildobby) [reply models/dex/dex_sandwiches.sql]: 👍 changed it to the usual structure of chain model that join into a crosschain one

**Comment by hildobby** [comment]
I figured out how to handle the blockchain field and changed to a loop approach + chain specific models @jeff-dude

**Comment by jeff-dude** [comment]
> I figured out how to handle the blockchain field and changed to a loop approach + chain specific models @jeff-dude
> 
> code it failing due to CI tests timing out but should work fine otherwise

thank you for the quick enhancements. i'm still a bit concerned on the timeouts. each of these read massive sets of data (from `dex.trades`) and then also joins in even larger data sets back to base tables. then that combined dataset is joined to itself multiple times in the same script.

how can we segment this further? or how can we take advantage of `dex.trades` already joining back to base tables previously? ideally we'd see these run in the allotted time. we've made it happen before on other large spells you've built. i can't help but think we need to dig deeper here

**Comment by hildobby** [comment]
> thank you for the quick enhancements. i'm still a bit concerned on the timeouts. each of these read massive sets of data (from `dex.trades`) and then also joins in even larger data sets back to base tables. then that combined dataset is joined to itself multiple times in the same script.
> 
> how can we segment this further? or how can we take advantage of `dex.trades` already joining back to base tables previously? ideally we'd see these run in the allotted time. we've made it happen before on other large spells you've built. i can't help but think we need to dig deeper here

Removed one distinct in the first CTE (it might work fine without it) and one join in `sandwiches` (the one that ensures there was 1+ trades sandwiched, so less info now but core still there, and includes failed MEV attempts) in each spell. 

Just as an explanation:
- First `trades` CTE needs to be grouped (some tokens have buy/sell limits but they are circumvented by having multiple trades of buy/sells in the same pool in the same tx, and I want the total traded in that case)
- The `sandwiches` CTE ensures that there is a frontrun & a backrun trade in the same block
- last join is also necessary as it ensures there aren't any duplicate entries as those happen if multiple frontrun or backrun txs are present (it happens)

Not sure how else I can better optimise it here but imo should already be a lot better, lmk if you have other suggestions

**Comment by jeff-dude** [comment]
> > thank you for the quick enhancements. i'm still a bit concerned on the timeouts. each of these read massive sets of data (from `dex.trades`) and then also joins in even larger data sets back to base tables. then that combined dataset is joined to itself multiple times in the same script.
> > how can we segment this further? or how can we take advantage of `dex.trades` already joining back to base tables previously? ideally we'd see these run in the allotted time. we've made it happen before on other large spells you've built. i can't help but think we need to dig deeper here
> 
> Removed one distinct in the first CTE (it might work fine without it) and one join in `sandwiches` (the one that ensures there was 1+ trades sandwiched, so less info now but core still there, and includes failed MEV attempts) in each spell.
> 
> Just as an explanation:
> 
> * First `trades` CTE needs to be grouped (some tokens have buy/sell limits but they are circumvented by having multiple trades of buy/sells in the same pool in the same tx, and I want the total traded in that case)
> * The `sandwiches` CTE ensures that there is a frontrun & a backrun trade in the same block
> * last join is also necessary as it ensures there aren't any duplicate entries as those happen if multiple frontrun or backrun txs are present (it happens)
> 
> Not sure how else I can better optimise it here but imo should already be a lot better, lmk if you have other suggestions

might be worth experimenting with separating by blockchain here too. for testing, can you do just one blockchain and save rest for later? i'd like to see if one running at once helps

File: .github/workflows/dbt_slim_ci.yml [hunk 2]
```diff
@@ -17,7 +17,7 @@ concurrency:
 jobs:
   dbt-test:
     runs-on: [ self-hosted, linux, spellbook ]
-    timeout-minutes: 90
+    timeout-minutes: 300
```
Comment (jeff-dude) [inline .github/workflows/dbt_slim_ci.yml:20]: we can revert this back. it's okay if it times out next run.

File: models/dex/arbitrum/dex_arbitrum_sandwiches.sql [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    , t.index
+    , MIN(t.gas_price) AS gas_price
+    , MIN((t.gas_price/POWER(10, 18))*t.gas_used) AS tx_fee
+    , SUM(COALESCE(dt.token_sold_amount_raw, 0)) AS token_sold_amount_raw
+    , SUM(COALESCE(dt.token_bought_amount_raw, 0)) AS token_bought_amount_raw
+    , SUM(COALESCE(dt.token_sold_amount, 0)) AS token_sold_amount
+    , SUM(COALESCE(dt.token_bought_amount, 0)) AS token_bought_amount
+    , SUM(COALESCE(dt.amount_usd, 0)) AS amount_usd
+    FROM {{ ref('dex_trades') }} dt
+    INNER JOIN {{ source('arbitrum', 'transactions') }} t ON t.block_time=dt.block_time
```
Comment (jeff-dude) [inline models/dex/arbitrum/dex_arbitrum_sandwiches.sql:40]: similar to recent PR, can you add the incremental filters on joins to base tables? it's a good general rule of thumb for all joins to tables with event-driven or time series data (i.e. base, prices, most decoded)

Comment (jeff-dude) [reply models/dex/arbitrum/dex_arbitrum_sandwiches.sql:40]: same idea for all blockchains

**Review by jeff-dude — CHANGES_REQUESTED** [review]
few minor comments below, then we should be good to go

**Pull request merged**